### ISSUE TITLE
Consider ClangFormat support to allow reformatting

### ISSUE THREAD

**Issue by ghost** [openframeworks/openFrameworks#3392]
Split out of #3390.

If we were to codify the OF style
  https://github.com/openframeworks/openFrameworks/tree/master/scripts/dev/style
  https://github.com/openframeworks/openFrameworks/wiki/openFrameworks-Coding-style-guidelines
  https://github.com/openframeworks/openFrameworks/wiki/oF-code-style

into a `.clang-format` file at the root of the repo, we could use [ClangFormat](http://clang.llvm.org/docs/ClangFormat.html) to reformat selected code to the OF style.

**Comment by pizthewiz** [comment]
Related forum thread: [Step-by-step guide to using Uncrustify in Code::Blocks](http://forum.openframeworks.cc/t/step-by-step-guide-to-using-uncrustify-in-code-blocks/18206)

**Comment by LeoColomb** [comment]
> I welcome everything that makes it easier for users to follow our style.

Could we start by adding the `.editorconfig` file? Easy and soft.
See http://editorconfig.org/.

**Comment by bilderbuchi** [comment]
The tiny [list of supported properties](https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties#current-universal-properties) of Editorconfig maybe make it possible to specify about 10%(?) of the code style (basically tabs-or-spaces, and their widths). Kindly compare this to the [uncrustify config file](https://github.com/openframeworks/openFrameworks/blob/master/scripts/dev/style/openFrameworks_style.cfg) which codifies the current OF code style.

**Comment by bilderbuchi** [comment]
Also, are you currently using an .editorconfig file in OF (e.g. in any directory above the repo, judging by the [lookup rules](http://editorconfig.org/#file-location), and how is it working for you to enforce style?

**Comment by LeoColomb** [comment]
> Editorconfig maybe make it possible to specify about 10%(?)

Yep, it's a first step.

> Kindly compare this to the uncrustify config file which codifies the current OF code style.

Yep. :laughing: But not applied yet, is it?

> are you currently using an .editorconfig file in OF

No: even without, VS (and not only 2015) complain about mixed style in every files of the oF core.
But I use it for all other projects (oF apps & other), and it's very useful, well managed and configurable.

> and how is it working for you to enforce style?

This is the good point: directly in IDE. [There are plugins](http://editorconfig.org/#download).
OK, plugins have to be installed, but each file' edition is fixed before saving.

> in any directory above the repo

The closer `.editorconfig` in the folder hierarchy apply its configs above the 2nd level `.editorconfig`, etc, up to the `.editorconfig` which include `root=true`.

**Comment by bilderbuchi** [comment]
> each file' edition is fixed before saving.

how can you consider a file "fixed" if you can't even tell the config file how the style should look like? :confused: 

> The closer .editorconfig in the folder hierarchy apply its configs above the 2nd level .editorconfig, etc, up to the .editorconfig which include root=true.

Yes, I know. this means, even if OF does not have such a file, you can easily place one in any dir above the OF repo, and this will get used on the OF code you edit. So, an immediate workaround to your problem.

**Comment by bilderbuchi** [comment]
> But not applied yet, is it?

If you mean that the code is not yet homogeneously formatted, then yes. The script is functional, though, and some people use it to format their fresh contributions.

**Comment by LeoColomb** [comment]
Yes, no, of course, I mean, it's a way to generalize #3390 to all IDE.

**Comment by flvi0** [comment]
Just want to share my personal opinion on how I think one should go about this. 
As @bilderbuchi mentioned in #4448, I share the @openframework/core opinion that having an automated script doing this upon checkin will heavily pollute the git history. The reason for that is that the git commits are hashed, and if you would automatically adjust the format when someone merges a branch, for example, you would necessarily have to introduced an extra commit as soon as a single character is changed by this.
Having an automated checker in the CI system is imo also not the right way to go about this. Since you have defined a rather rigorous style document (which is not bad in itself, one might argue about this), people will quickly be put off by this enforcement if they didn't get it quite right for the second or third time. I think this will make people refrain from commiting to the public repo at all, and this is something which should be avoided at all cost.
Another aspect is that it is good practice when working with git, to clearly separate into distinct commits any functional changes from cosmetics / reformatting. Otherwise it becomes difficullt to impossible (i.e. would require costly conflict resolution by hand) if people want to merge only specific feature sets in given branches, e.g. for backporting bugfixes into release branches, or allowing people to truly work on distinct development branches and merge back and forth / cherrypick between them. 
This is why, as I intended to do with the one-pass gnu indent approach (or tool of your choice, seems to uncrustify), in my opinion it makes sense to just do a single one-pass reformatting of the whole tree *now and then*. This will not put people off commiting, create only a single commit in history for all the cosmetics, and will allow people to concentrate on functional changes with their actual commits. A bit care should be taken to make those changes at some point in time before some important branches diverge, and not thereafter, e.g. if you want to backport bugfixes to 0.9.x for some time, it should be done before 0.9 branches off, otherwise you will of course have the conflicts.
A makefile/xcode/vs (cmake anyone?) build target could be supplied, which will just be run over the whole project, i.e. specifically the openFrameworks source directory, no need to bother with accidentally reformatting the other libs you deploy alongside oF (which is another issue). This could quickly be done by some core maintainer as soon as it is seen fit. Any 'degeneration' this not-so-strict enforcement could cause I don't see, since you can just refrain from merging a branch if it looks (too) ill-formatted to you and kindly point the contributor to adhere to the style better and refine. 
This middle-ground would be the best solution (in my conception that is) and that's why I earlier though let me just do this quickly in order to move on to more interesting topics.

Best

**Comment by bilderbuchi** [comment]
> people will quickly be put off by this enforcement if they didn't get it quite righ

the idea was that the automated checker would report back to the PR if it found style fixes and offer an easy-to-apply diff (e.g. via a gist) that the submitter can just add to the PR painlessly, and also instructions how to to it yourself.

> run over the whole project, i.e. specifically the openFrameworks source directory, no need to bother with accidentally reformatting the other libs you deploy alongside oF (which is another issue)

this is what `styleWholeOF` does.

> Having an automated checker in the CI system is imo also not the right way [...] people will quickly be put off by this enforcement [...]

> since you can just refrain from merging a branch if it looks (too) ill-formatted to you and kindly point the contributor to adhere to the style better and refine. 

erm, I think you just switched positions - the automated checker would be exactly for doing this kind pointing out... :-)

**Comment by flvi0** [comment]
> the idea was that the automated checker would report back to the PR if it found style fixes and offer an easy-to-apply diff (e.g. via a gist) that the submitter can just add to the PR painlessly

Yes, but having this automated and always enforced will again pollute the git history if each PR ends with a 'formatting commit'. Also I think you didn't get my point about separating functional from syntactic changes, since (I suppose) that suggested diff would apply the formatting to the whole file, touching lines which are not part of the functional changes. Thus my suggestion was to take the approach to rather allow some "slack" in commits (PR's might still be rejected if they are *too* badly formatted), and just fix the whole project's formatting once in a while with a single commit.

**Comment by flvi0** [comment]
One could argue that when merging the PR to different (e.g. dev/release) branches, on could just leave out the last formatting commit or such, but I believe that won't scale up and unnecessarily complicate things.

**Comment by bilderbuchi** [comment]
> since (I suppose) that suggested diff would apply the formatting to the whole file, touching lines which are not part of the functional changes.

When the code is kept in-style with an automated mechanism (which has ben the working hypothesis), this will not have an impact, as those other lines will conform to the style and thus won't be affected.  
Of course, doing formatting sweeps from time to time is just as viable a strategy.

Our branching strategy is not that complicated (basically, just a `master` branch for development, and `stable` always points to the latest release, + feature branches for PRs), so I think accomodating formatting commits won't be a problem.

**Comment by flvi0** [comment]
Well yeah true of course, so we'd do an initial sweep, then have people automatically noticed and the diff/patch sent. They could even be told by the bot to apply the diff via a commit --amend s.t. there will be no spurious history entries. Anyways I think one should not spend too large amounts of time on this, and a sweep not and then is a cheap solution which (I think) will do just as well.
